app/gui2/src/util/ast/__tests__/aliasAnalysis.test.ts

```diff
@@ -0,0 +1,319 @@
+import type { ContentRange } from '../../../../shared/yjsModel'
+import { AliasAnalyzer } from '../aliasAnalysis'
+
+/** The type of annotation. */
+enum AnnotationType {
+  /** An identifier binding (introducing variable). */
+  Binding = 'Binding',
+  /** An identifier usage. */
+  Usage = 'Usage',
+}
+
+/** Information about an annotated identifier. */
+class Annotation {
+  /**
+   * @param kind Whether this is an identifier binding or usage.
+   * @param id The special user-defined id to disambiguate identifiers with the same name.
+   */
+  constructor(
+    public kind: AnnotationType,
+    public id: number,
+  ) {}
+}
+
+/** Parse annotations from the annotated code. See the file-top comment for the syntax. */
+function parseAnnotations(annotatedCode: string): {
+  unannotatedCode: string
+  annotations: ObjectKeyedMap<ContentRange, Annotation>
+} {
+  const annotations = new ObjectKeyedMap()
+
+  // Iterate over all annotations (either bindings or usages).
+  // I.e. we want to cover both `«1,x»` and `»1,x«` cases, while keeping the track of the annotation type.
+  const annotationRegex = /«(\d+),([^«»]+)»|»(\d+),([^«»]+)«/g
+
+  // As the annotations are removed from the code, we need to keep track of the offset between the annotated and
+  // unannotated code. This is necessary to correctly calculate the start and length of the annotations.
+  let accumulatedOffset = 0
+
+  const unannotatedCode = annotatedCode.replace(
+    annotationRegex,
+    (match, bindingPrefix, bindingName, usagePrefix, usageName, offset) => {
+      console.log(`Processing annotated identifier ${match}.`)
+
+      // Sanity check: either both binding prefix and name are present, or both usage prefix and name are present.
+      // Otherwise, we have an internal error in the regex.
+      assertEqual(bindingPrefix != null, bindingName != null)
+      assertEqual(usagePrefix != null, usageName != null)
+      assertNotEqual(bindingPrefix != null, usagePrefix != null)
+
+      const id = parseInt(bindingPrefix ?? usagePrefix, 10)
+      const name = bindingName ?? usageName
+      const kind = bindingPrefix != null ? AnnotationType.Binding : AnnotationType.Usage
+
+      const start = offset - accumulatedOffset
+      const end = start + name.length
+      const range = [start, end]
+
+      const annotation = new Annotation(kind, id)
+      accumulatedOffset += match.length - name.length
+      annotations.set(range, annotation)
+      return name
+    },
+  )
+  // Make sure that all annotations were removed. i.e. make sure that there are no unmatched `«` or `»` characters.
+  assert(!unannotatedCode.includes('«'), `Unmatched « character. Full code: ${unannotatedCode}`)
+  assert(!unannotatedCode.includes('»'), `Unmatched » character. Full code: ${unannotatedCode}`)
+  return { unannotatedCode, annotations }
+}
+
+/** Alias analysis test case, typically parsed from an annotated code. */
+class TestCase {
```

Comment:
I don't see why this needs to be a class. Looks like a simple function would be totally enough here.

Reply:
This is for convenience, as we have a few functions that want to work with the same data (and, in general, I find method notation more readable when compared to free functions)